File: crates/brainwires-core/src/provider.rs
use anyhow::Result;
use async_trait::async_trait;
use futures::stream::BoxStream;
use serde::{Deserialize, Serialize};

use crate::message::{ChatResponse, Message, StreamChunk};
use crate::tool::Tool;

/// Base provider trait for AI providers
#[async_trait]
pub trait Provider: Send + Sync {
    /// Get the provider name
    fn name(&self) -> &str;

    /// Get the model's maximum output tokens (for setting appropriate limits)
    /// Returns None if the model doesn't have a specific limit
    fn max_output_tokens(&self) -> Option<u32> {
        None // Default implementation - providers can override
    }

    /// Chat completion (non-streaming)
    async fn chat(
        &self,
        messages: &[Message],
        tools: Option<&[Tool]>,
        options: &ChatOptions,
    ) -> Result<ChatResponse>;

    /// Chat completion (streaming)
    fn stream_chat<'a>(
        &'a self,
        messages: &'a [Message],
        tools: Option<&'a [Tool]>,
        options: &'a ChatOptions,
    ) -> BoxStream<'a, Result<StreamChunk>>;
}

/// Prompt-cache strategy for providers that support explicit caching
/// (Anthropic Messages API today; a no-op elsewhere).
///
/// Controls which parts of a request receive `cache_control` breakpoints.
/// Caching reuses cached prompt bytes across turns for a 50–90% input-token
/// discount on subsequent calls, at the cost of a one-time "creation" charge
/// on first population.
#[derive(Debug, Clone, Copy, Default, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum CacheStrategy {
    /// No cache breakpoints. Fresh compute on every call.
    Off,
    /// Cache only the system prompt.
    SystemOnly,
    /// Cache the system prompt and tool definitions (the default).
    #[default]
    SystemAndTools,
    /// Cache system + tools + the tail of the conversation once the message
    /// history reaches the given approximate token threshold.
    SystemAndTailTurn {
        /// Minimum conversation size (approximate tokens) before the tail
        /// breakpoint is emitted. Avoids wasting a cache slot on short chats.
        threshold_tokens: u32,
    },
}

/// Chat completion options
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ChatOptions {
    /// Temperature (0.0 - 1.0)
    #[serde(skip_serializing_if = "Option::is_none")]
    pub temperature: Option<f32>,
    /// Maximum tokens to generate
    #[serde(skip_serializing_if = "Option::is_none")]
    pub max_tokens: Option<u32>,
    /// Top-p sampling
    #[serde(skip_serializing_if = "Option::is_none")]
    pub top_p: Option<f32>,
    /// Stop sequences
    #[serde(skip_serializing_if = "Option::is_none")]
    pub stop: Option<Vec<String>>,
    /// System prompt
    #[serde(skip_serializing_if = "Option::is_none")]
    pub system: Option<String>,
    /// Per-request model override.
    ///
    /// When `Some`, providers MUST use this model name instead of their default.
    /// This enables per-session model switching without replacing the provider.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub model: Option<String>,
    /// Prompt-cache strategy. Ignored by providers without prompt caching.
    #[serde(default)]
    pub cache_strategy: CacheStrategy,
}

impl Default for ChatOptions {
    fn default() -> Self {
        Self {
            temperature: Some(0.7),
            max_tokens: Some(4096),
            top_p: None,
            stop: None,
            system: None,
            model: None,
            cache_strategy: CacheStrategy::default(),
        }
    }
}

impl ChatOptions {
    /// Create new chat options with defaults
    pub fn new() -> Self {
        Self::default()
    }

    /// Set temperature
    pub fn temperature(mut self, temperature: f32) -> Self {
        self.temperature = Some(temperature);
        self
    }

    /// Set max tokens
    pub fn max_tokens(mut self, max_tokens: u32) -> Self {
        self.max_tokens = Some(max_tokens);
        self
    }

    /// Set system prompt
    pub fn system<S: Into<String>>(mut self, system: S) -> Self {
        self.system = Some(system.into());
        self
    }

    /// Set top-p sampling
    pub fn top_p(mut self, top_p: f32) -> Self {
        self.top_p = Some(top_p);
        self
    }

    /// Override the model for this request.
    pub fn model<S: Into<String>>(mut self, model: S) -> Self {
        self.model = Some(model.into());
        self
    }

    /// Set the prompt-cache strategy.
    pub fn cache_strategy(mut self, strategy: CacheStrategy) -> Self {
        self.cache_strategy = strategy;
        self
    }

    /// Deterministic classification/routing (temp=0, few tokens)
    pub fn deterministic(max_tokens: u32) -> Self {
        Self {
            temperature: Some(0.0),
            max_tokens: Some(max_tokens),
            ..Default::default()
        }
    }

    /// Low-temperature factual generation
    pub fn factual(max_tokens: u32) -> Self {
        Self {
            temperature: Some(0.1),
            max_tokens: Some(max_tokens),
            top_p: Some(0.9),
            ..Default::default()
        }
    }

    /// Creative generation with moderate temperature
    pub fn creative(max_tokens: u32) -> Self {
        Self {
            temperature: Some(0.3),
            max_tokens: Some(max_tokens),
            ..Default::default()
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_chat_options_default() {
        let opts = ChatOptions::default();
        assert_eq!(opts.temperature, Some(0.7));
        assert_eq!(opts.max_tokens, Some(4096));
    }

    #[test]
    fn test_chat_options_builder() {
        let opts = ChatOptions::new()
            .temperature(0.5)
            .max_tokens(2048)
            .system("Test");
        assert_eq!(opts.temperature, Some(0.5));
        assert_eq!(opts.max_tokens, Some(2048));
        assert_eq!(opts.system, Some("Test".to_string()));
    }

    #[test]
    fn test_chat_options_deterministic() {
        let opts = ChatOptions::deterministic(50);
        assert_eq!(opts.temperature, Some(0.0));
        assert_eq!(opts.max_tokens, Some(50));
    }

    #[test]
    fn test_chat_options_factual() {
        let opts = ChatOptions::factual(200);
        assert_eq!(opts.temperature, Some(0.1));
        assert_eq!(opts.max_tokens, Some(200));
        assert_eq!(opts.top_p, Some(0.9));
    }

    #[test]
    fn test_chat_options_creative() {
        let opts = ChatOptions::creative(400);
        assert_eq!(opts.temperature, Some(0.3));
        assert_eq!(opts.max_tokens, Some(400));
    }
}

File: deno/packages/core/provider.ts
import type { ChatResponse, Message, StreamChunk } from "./message.ts";
import type { Tool } from "./tool.ts";

/** Base provider interface for AI providers.
 * Equivalent to Rust's `Provider` trait in rullama-core. */
export interface Provider {
  /** Get the provider name. */
  readonly name: string;

  /** Get the model's maximum output tokens. Returns undefined if no specific limit. */
  maxOutputTokens?(): number | undefined;

  /** Chat completion (non-streaming). */
  chat(
    messages: Message[],
    tools: Tool[] | undefined,
    options: ChatOptions,
  ): Promise<ChatResponse>;

  /** Chat completion (streaming). */
  streamChat(
    messages: Message[],
    tools: Tool[] | undefined,
    options: ChatOptions,
  ): AsyncIterable<StreamChunk>;
}

/** Chat completion options.
 * Equivalent to Rust's `ChatOptions` in rullama-core. */
export class ChatOptions {
  /** Sampling temperature (default 0.7). */
  temperature?: number;
  /** Maximum tokens the model may generate (default 4096). */
  max_tokens?: number;
  /** Nucleus-sampling probability mass. */
  top_p?: number;
  /** Sequences that end generation when produced. */
  stop?: string[];
  /** System prompt sent ahead of the messages. */
  system?: string;
  /**
   * Model to use for this call, when the provider serves several. Providers
   * bound to one model ignore it; decorators (circuit breaker, cache) key
   * their state per model with it.
   */
  model?: string;

  /** Create options, applying the defaults (`temperature` 0.7, `max_tokens` 4096) for anything not given. */
  constructor(opts?: Partial<ChatOptions>) {
    this.temperature = opts?.temperature ?? 0.7;
    this.max_tokens = opts?.max_tokens ?? 4096;
    this.top_p = opts?.top_p;
    this.stop = opts?.stop;
    this.system = opts?.system;
    this.model = opts?.model;
  }

  /** Set the model (builder). */
  setModel(model: string): this {
    this.model = model;
    return this;
  }

  /** Create new chat options with defaults. */
  static create(): ChatOptions {
    return new ChatOptions();
  }

  /** Set temperature (builder). */
  setTemperature(temperature: number): this {
    this.temperature = temperature;
    return this;
  }

  /** Set max tokens (builder). */
  setMaxTokens(maxTokens: number): this {
    this.max_tokens = maxTokens;
    return this;
  }

  /** Set system prompt (builder). */
  setSystem(system: string): this {
    this.system = system;
    return this;
  }

  /** Set top-p sampling (builder). */
  setTopP(topP: number): this {
    this.top_p = topP;
    return this;
  }

  /** Deterministic classification/routing (temp=0, few tokens). */
  static deterministic(maxTokens: number): ChatOptions {
    return new ChatOptions({ temperature: 0.0, max_tokens: maxTokens });
  }

  /** Low-temperature factual generation. */
  static factual(maxTokens: number): ChatOptions {
    return new ChatOptions({
      temperature: 0.1,
      max_tokens: maxTokens,
      top_p: 0.9,
    });
  }

  /** Creative generation with moderate temperature. */
  static creative(maxTokens: number): ChatOptions {
    return new ChatOptions({ temperature: 0.3, max_tokens: maxTokens });
  }

  /** Serialize to JSON, omitting undefined fields. */
  toJSON(): Record<string, unknown> {
    const obj: Record<string, unknown> = {};
    if (this.temperature !== undefined) obj.temperature = this.temperature;
    if (this.max_tokens !== undefined) obj.max_tokens = this.max_tokens;
    if (this.top_p !== undefined) obj.top_p = this.top_p;
    if (this.stop !== undefined) obj.stop = this.stop;
    if (this.system !== undefined) obj.system = this.system;
    if (this.model !== undefined) obj.model = this.model;
    return obj;
  }
}
